File: app/src/main/java/org/metabrainz/mobile/presentation/features/barcode/BarcodeViewModel.java
package org.metabrainz.mobile.presentation.features.barcode;

import androidx.hilt.lifecycle.ViewModelInject;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import org.metabrainz.mobile.data.repository.BarcodeLookupRepository;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.Release;

import java.util.List;

public class BarcodeViewModel extends ViewModel {

    private BarcodeLookupRepository barcodeLookupRepository;
    private MutableLiveData<List<Release>> barcodeLiveData;

    @ViewModelInject
    public BarcodeViewModel(BarcodeLookupRepository repository) {
        this.barcodeLookupRepository = repository;
    }

    public MutableLiveData<List<Release>> getBarcodeLiveData() {
        if (barcodeLiveData == null) barcodeLiveData = barcodeLookupRepository.getReleaseListData();
        return barcodeLiveData;
    }

    public void fetchReleasesWithBarcode(String barcode) {
        barcodeLookupRepository.lookupReleasesWithBarcode("barcode:" + barcode);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        this.barcodeLookupRepository = null;
    }
}

File: app/src/main/java/org/metabrainz/mobile/presentation/features/barcode/BarcodeViewModel.kt
package org.metabrainz.mobile.presentation.features.barcode

import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.liveData
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers
import org.metabrainz.mobile.data.repository.BarcodeLookupRepository
import org.metabrainz.mobile.data.sources.api.entities.mbentity.Release
import org.metabrainz.mobile.util.Resource
import javax.inject.Inject

@HiltViewModel
class BarcodeViewModel @Inject constructor(private var barcodeLookupRepository: BarcodeLookupRepository) : ViewModel() {

    fun fetchReleasesWithBarcode(barcode: String): LiveData<Resource<List<Release>>> {
        return liveData(viewModelScope.coroutineContext + Dispatchers.IO) {
            emit(barcodeLookupRepository.lookupReleasesWithBarcode("barcode:$barcode"))
        }
    }

}